parquet-hadoop/src/main/java/org/apache/parquet/crypto/AesCipher.java

```diff
@@ -0,0 +1,128 @@
+  protected static final int CHUNK_LENGTH = 4 * 1024;
+  protected static final int SIZE_LENGTH = ModuleCipherFactory.SIZE_LENGTH;
+
+  protected EncryptionKey aesKey;
+  protected final SecureRandom randomGenerator;
+  protected Cipher cipher;
+  protected final byte[] localNonce;
+  protected boolean wipedOut;
+
+
+  AesCipher(AesMode mode, byte[] keyBytes) throws IllegalArgumentException, IOException {
+    if (null == keyBytes) {
+      throw new IllegalArgumentException("Null key bytes");
+    }
+
+    aesKey = new EncryptionKey(keyBytes);
+    randomGenerator = new SecureRandom();
+    localNonce = new byte[NONCE_LENGTH];
+    wipedOut = false;
+  }
+
+  public static byte[] createModuleAAD(byte[] fileAAD, ModuleType moduleType, 
+      short rowGroupOrdinal, short columnOrdinal, short pageOrdinal) {
+    byte[] typeOrdinalBytes = new byte[1];
+    typeOrdinalBytes[0] = moduleType.getValue();
+    if (ModuleType.Footer == moduleType) {
+      return concatByteArrays(fileAAD, typeOrdinalBytes);      
+    }
+
+    byte[] rowGroupOrdinalBytes = shortToBytesLE(rowGroupOrdinal);
+    byte[] columnOrdinalBytes = shortToBytesLE(columnOrdinal);
+    if (ModuleType.DataPage != moduleType && ModuleType.DataPageHeader != moduleType) {
+      return concatByteArrays(fileAAD, typeOrdinalBytes, rowGroupOrdinalBytes, columnOrdinalBytes); 
+    }
+
+    byte[] pageOrdinalBytes = shortToBytesLE(pageOrdinal);
+    return concatByteArrays(fileAAD, typeOrdinalBytes, rowGroupOrdinalBytes, columnOrdinalBytes, pageOrdinalBytes);
+  }
+
+  public static byte[] createFooterAAD(byte[] aadPrefixBytes) {
+    return createModuleAAD(aadPrefixBytes, ModuleType.Footer, (short) -1, (short) -1, (short) -1);
+  }
+
+  /**
+   * Update last two bytes with new page ordinal (instead of creating new page AAD from scratch)
+   * @param pageAAD
+   * @param newPageOrdinal
+   */
+  public static void quickUpdatePageAAD(byte[] pageAAD, short newPageOrdinal) {
+    byte[] pageOrdinalBytes = shortToBytesLE(newPageOrdinal);
+    int length = pageAAD.length;
+    System.arraycopy(pageOrdinalBytes, 0, pageAAD, length-2, 2);
+  }
+
+  static byte[] concatByteArrays(byte[]... arrays) {
+    int totalLength = 0;
+    for (byte[] array : arrays) {
+      totalLength += array.length;
+    }
+    byte[] output = new byte[totalLength];
+    int offset = 0;
+    for (byte[] array : arrays) {
+      int arrayLength = array.length;
+      System.arraycopy(array, 0, output, offset, arrayLength);
+      offset += arrayLength;
+    }
+    return output;
+  }
+
+  private static byte[] shortToBytesLE(short input) {
+    byte[] output  = new byte[2];
+    output[1] = (byte)(0xff & (input >> 8));
+    output[0] = (byte)(0xff & (input));
+    return output;
+  }
+
+  public void wipeOut() {
+    wipedOut = true;
+    try {
+      aesKey.destroy();
+    } catch (DestroyFailedException e) {
+      throw new ShouldNeverHappenException(e);
```

Comment:
In the case of Exception, do we consider wipeOut as true or false?

Reply:
good point, should be false. I'll move the flag change to the end.